### DISCUSSION TITLE
Importing external Python files during runtime fails at executing the imported module

### DISCUSSION THREAD

**Discussion by silverv** [pyinstaller/pyinstaller#7214] [Help]
This problem is very similar to: https://github.com/pyinstaller/pyinstaller/discussions/5546.
```
Traceback (most recent call last):
File "yukon\services\dtype_loader.py", line 62, in _load
mod = importlib.import_module(name)
File "importlib\__init__.py", line 126, in import_module
File "", line 1050, in _gcd_import
File "", line 1027, in _find_and_load
File "", line 1006, in _find_and_load_unlocked
File "", line 688, in _load_unlocked
File "", line 883, in exec_module
File "", line 241, in _call_with_frames_removed
File "C:\Users\silver\AppData\Local\Temp\_MEI151723\.compiled\reg\udral\physics\electricity\__init__.py", line 6, in
from reg.udral.physics.electricity.Power_0_1 import Power_0_1 as Power_0_1
File "C:\Users\silver\AppData\Local\Temp\_MEI151723\.compiled\reg\udral\physics\electricity\Power_0_1.py", line 21, in
import uavcan.si.unit.electric_current
ModuleNotFoundError: No module named 'uavcan.si'
```

This error occurs only when I build the executable (for Yukon ) with Pyinstaller, before the error happens, I log the current sys.path as such: 
```
221102_163851 dtype_loader Current PYTHONPATH: ['C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\base_library.zip', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\lib-dynload', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723', 'C:\\Users\\silver\\.pycyphal', 'C:\\Users\\silver\\.pycyphal\\zubax', 'C:\\Users\\silver\\.pycyphal\\uavcan\\si', 'C:\\Users\\silver\\.pycyphal\\uavcan', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\reg', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\sirius_cyber_corp', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\uavcan', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\zubax', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\zubax_internet', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\reg', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\sirius_cyber_corp', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\uavcan', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\zubax', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\zubax_internet', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\reg', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\sirius_cyber_corp', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\uavcan', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\zubax', 'C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled\\zubax_internet']
```
It contains `C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled` and `C:\\Users\\silver\\.pycyphal`, both of which are folders which contain identical uavcan packages. Only one would be necessary usually but I've tried to play it safe in this case and put as many thing in sys.path as possible.

Why does the Python in my executable (create by Pyinstaller) not want to import these packages? Or why can it import uavcan but not uavcan.si? uavcan.si is a package inside the uavcan package. To experiment with things I have also added the uavcan package folder path to sys.path but that didn't help.

**Comment by rokm** [comment]
> `mod = importlib.import_module(name)` 

PyInstaller automatically picks up only `import modname` types of imports; it will not pick up an import made via `importlib.import_module(modname)` (nor `__import__(modname)`). To deal with such cases, you need to add the module to hidden imports, e.g., add [`--hiddenimport uavcan.si`](https://pyinstaller.org/en/stable/usage.html#cmdoption-hidden-import) to your PyInstaller command. If there are many indirect imports in the package, it might be easier to just automatically collect all submodules by adding [`--collect-submodules uavcan`](https://pyinstaller.org/en/stable/usage.html#cmdoption-collect-submodules) (or `--collect-submodules uavcan.si`, if only a sub-package is problematic).

**Reply by rokm** [reply]
Once you have hidden imports sorted out, there should be no need to mess with `sys.path` (which in your case probably did more harm than good, considering you added both the directory containing packages and individual packages' directories to it...)

**Reply by silverv** [reply]
The directories that are going to be imported are not known in advance

**Reply by silverv** [reply]
Can I work around this by using exec("import uavcan.si")?

**Reply by silverv** [reply]
The hidden imports are okay, there's no problem there.
https://github.com/OpenCyphal-Garage/yukon/blob/037df8bd2217e3c1466142ef25a8c4f6750a66fa/pyinstaller.spec#L98

**Reply by rokm** [reply]
Can you clean up your `sys.path` then to contain only directories that provide packages (and not package directories themselves), and also clean it up so that only one directory contains a copy of a package?

**Reply by rokm** [reply]
So from those extra directories, either only `C:\\Users\\silver\\AppData\\Local\\Temp\\_MEI151723\\.compiled` (i.e.,  `os.path.join(sys._MEIPASS, '.compiled')`) or `C:\\Users\\silver\\.pycyphal`  should be in `sys.path`. EDIT: or both, since python probably takes the package from the first location it finds.

**Reply by rokm** [reply]
And at the same time, check your `build/<name>/PYZ-00.toc` to see if any `uavcan` modules are also collected in PYZ. Could be that a partially-collected package in PYZ is preventing the on-disk ones from being discovered.

**Reply by silverv** [reply]
In my case it's fine if either one gets discovered, that's the reason I've left it there, but I'll remove the excessive entries from path again (did that as an experiment after reading the above mentioned post where it was mentioned that Pyinstaller can literally only load packages in path.)

**Reply by rokm** [reply]
> In my case it's fine if either one gets discovered, that's the reason I've left it there, but I'll remove the excessive entries from path again (did that as an experiment after reading the above mentioned post where it was mentioned that Pyinstaller can literally only load packages in path.)

Yes, that's true. But if you put individual package directories in `sys.path` (as opposed to their parent directory, e.g. `C:\\Users\\silver\\.pycyphal`), then the modules of the package become top-level modules, which may cause conflicts with actual top-level modules.

But packages collected into executable-embedded PYZ archive have precedence when being loaded, so if there is a partially-collected `uavcan` package in PYZ, the on-disk `uavcan` packages in `sys.path` won't be discovered (I think the only exemption to that is if you put `uavcan` directory directly into top-level application directory (instead of `.compiled` subdirectory, because the package in PYZ is pretending to be physically located in top-level directory). I think the best way to solve this is to collect whole `uavcan` into PYZ (using `collect_submodules`) and drop the local copies; or alternatively, exclude the `uavcan` (so it is not collected into PYZ at all) and therefore force the use of external copies.

**Reply by silverv** [reply]
This last comment is the answer! Thanks. uavcan was partially included in a pyz file because it was imported and then exactly as you said.
